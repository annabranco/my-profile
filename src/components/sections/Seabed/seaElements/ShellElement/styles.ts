import styled, { keyframes } from 'styled-components';

export const TheShell = styled.img`
  width: 60px;
  opacity: 0.5;
  filter: drop-shadow(0 5px 2px rgba(0, 0, 0, 0.2));
`;
TheShell.displayName = 'TheShell';

export const ShellContainer = styled.div`
  z-index: 100;
  position: absolute;
  bottom: 30px;
  left: 15vw;
`;
ShellContainer.displayName = 'ShellContainer';

export const PearlShine = styled.div`
  position: absolute;
  top: 1px;
  left: 1px;
  height: 8px;
  width: 8px;
  border-radius: 50%;
  background: radial-gradient(circle at 100%, rgb(220, 220, 220), white);
  box-shadow: 0 0 10px 1px white;
`;
PearlShine.displayName = 'PearlShine';

const Shine = keyframes`
0% {
    box-shadow: 0 0 20px 4px yellow;
  }

  50% {
        box-shadow: none;
  }

  100% {
        box-shadow: 0 0 20px 4px yellow;
  }
`;

export const Pearl = styled.div`
  position: absolute;
  bottom: 14px;
  left: 23px;
  height: 15px;
  width: 15px;
  background: radial-gradient(
    circle at 100%,
    rgb(220, 220, 220),
    rgb(192, 192, 192) 50%,
    rgb(255, 255, 255) 75%,
    rgb(220, 220, 220) 15%
  );
  border-radius: 50%;
  filter: drop-shadow(0 5px 2px rgba(0, 0, 0, 0.2));
  border: 1px solid rgba(92, 92, 92, 0.3);

  animation-name: ${Shine};
  animation-duration: 2s;
  animation-timing-function: linear;
  animation-iteration-count: infinite;
  animation-fill-mode: forwards;
`;
Pearl.displayName = 'Pearl';
