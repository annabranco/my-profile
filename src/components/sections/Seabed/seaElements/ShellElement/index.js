import React, { useState } from 'react';
import { bool, func } from 'prop-types';
import { ClosedShell, OpenedShell } from '../../../../../assets/images';
import FoundPearl from '../../../../../assets/sounds/found.mp3';
import Bong from '../../../../../assets/sounds/bong.mp3';

import { TheShell, ShellContainer, Pearl, PearlShine } from './styles';

const FindingSound = new Audio(FoundPearl);
const GettingSound = new Audio(Bong);

const ShellElement = ({ toggleHasPearl, hidden }) => {
  const [opened, open] = useState(false);
  const [hasPearl, getPearl] = useState(false);

  const onClickShell = () => {
    if (!opened) {
      open(true);
      FindingSound.currentTime = 0;
      FindingSound.volume = 0.5;
      FindingSound.play();
    }
  };

  const onClickPearl = () => {
    getPearl(true);
    GettingSound.currentTime = 0;
    GettingSound.volume = 0.1;
    GettingSound.play();
    toggleHasPearl(true);
  };

  return (
    <ShellContainer hidden={hidden}>
      <TheShell
        onClick={onClickShell}
        src={opened ? OpenedShell : ClosedShell}
      />
      {opened && !hasPearl && (
        <Pearl onClick={onClickPearl}>
          <PearlShine />
        </Pearl>
      )}
    </ShellContainer>
  );
};

ShellElement.propTypes = {
  toggleHasPearl: func.isRequired,
  hidden: bool.isRequired
};

export default ShellElement;
