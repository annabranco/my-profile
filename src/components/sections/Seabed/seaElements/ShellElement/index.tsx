import React, { Dispatch, ReactElement, SetStateAction } from 'react';
import { bool, func } from 'prop-types';
import { useStateWithLabel } from '../../../../../utils/hooks';
import { ClosedShell, OpenedShell } from '../../../../../assets/images';
import FoundPearl from '../../../../../assets/sounds/found.mp3';
import Bong from '../../../../../assets/sounds/bong.mp3';
import { TheShell, ShellContainer, Pearl, PearlShine } from './styles';

const FindingSound = new Audio(FoundPearl);
const GettingSound = new Audio(Bong);

interface Props {
  hidden: boolean;
  toggleHasPearl: Dispatch<SetStateAction<boolean>>;
}

const ShellElement = ({ toggleHasPearl, hidden }: Props): ReactElement => {
  const [opened, open] = useStateWithLabel<boolean>(false, 'pearlOpened');
  const [hasPearl, getPearl] = useStateWithLabel<boolean>(false, 'hasPearl');

  const onClickShell = (): void => {
    if (!opened) {
      open(true);
      FindingSound.currentTime = 0;
      FindingSound.volume = 0.5;
      FindingSound.play();
    }
  };

  const onClickPearl = (): void => {
    getPearl(true);
    GettingSound.currentTime = 0;
    GettingSound.volume = 0.1;
    GettingSound.play();
    toggleHasPearl(true);
  };

  return (
    <ShellContainer hidden={hidden}>
      <TheShell
        onClick={onClickShell}
        src={opened ? OpenedShell : ClosedShell}
      />
      {opened && !hasPearl && (
        <Pearl onClick={onClickPearl}>
          <PearlShine />
        </Pearl>
      )}
    </ShellContainer>
  );
};

ShellElement.propTypes = {
  hidden: bool.isRequired,
  toggleHasPearl: func.isRequired
};

export default ShellElement;
