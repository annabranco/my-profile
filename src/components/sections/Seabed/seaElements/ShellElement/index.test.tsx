import { ShallowWrapper, shallow } from 'enzyme';
import 'jest-styled-components';
import ShellElement from '.';

const callbackMock = jest.fn();

describe('< Shell Component >', () => {
  let wrapper: ShallowWrapper;

  beforeAll(() => {
    wrapper = shallow(
      <ShellElement toggleHasPearl={callbackMock} hidden={false} />
    );
    window.HTMLMediaElement.prototype.load = () => null;
    window.HTMLMediaElement.prototype.play = () =>
      new Promise(resolve => resolve());
  });

  it('should mount', () => {
    expect(wrapper).toBeDefined();
  });

  it('matches the snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have a Shell image element', () => {
    expect(wrapper.find('TheShell')).toHaveLength(1);
  });

  it('should have a Pearl image only displayed after Shell is clicked', () => {
    const shell = wrapper.find('TheShell');
    let pearl = wrapper.find('Pearl');
    expect(pearl).toHaveLength(0);

    shell.simulate('click');
    pearl = wrapper.find('Pearl');
    expect(pearl).toHaveLength(1);
  });

  it('should hide Pearl and trigger toggleHasPearl event when Pearl is clicked', () => {
    let pearl;
    wrapper.find('TheShell').simulate('click');
    pearl = wrapper.find('Pearl');
    expect(pearl).toHaveLength(1);

    pearl.simulate('click');
    pearl = wrapper.find('Pearl');
    expect(pearl).toHaveLength(0);
    expect(callbackMock).toHaveBeenCalledTimes(1);
  });
});
