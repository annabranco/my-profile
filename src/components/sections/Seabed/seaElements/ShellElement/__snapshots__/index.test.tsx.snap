// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`< Shell Component > matches the snapshot 1`] = `
<ShellContainer
  hidden={false}
>
  <TheShell
    onClick={[Function]}
    src={
      Object {
        "process": [Function],
      }
    }
  />
</ShellContainer>
`;
